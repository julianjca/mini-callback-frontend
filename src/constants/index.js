export const SET_LOGIN_STATE = 'SET_LOGIN_STATE'
export const SET_USER = 'SET_USER'
export const SET_AUTHENTICATING = 'SET_AUTHENTICATING'
export const DO_LOGOUT = 'DO_LOGOUT'
