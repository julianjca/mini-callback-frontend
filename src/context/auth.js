import React, { useEffect, createContext, useReducer } from 'react'
import cookie from 'react-cookies'
import axios from 'axios'

import { SET_LOGIN_STATE, SET_USER, SET_AUTHENTICATING } from '../constants'

const AuthStateContext = createContext()
const AuthDispatchContext = createContext()

const reducer = (state, action) => {
  switch (action.type) {
    case SET_LOGIN_STATE: {
      return { ...state, isLoggedIn: action.isLoggedIn }
    }

    case SET_USER: {
      return { ...state, user: action.user }
    }

    case SET_AUTHENTICATING: {
      return { ...state, isAuthenticating: action.isAuthenticating }
    }

    default:
      console.log('error')
  }
}

const AuthProvider = ({ children }) => {
  const [state, dispatch] = useReducer(reducer, {
    isLoggedIn: false,
    user: null,
    isAuthenticating: true,
  })

  useEffect(() => {
    const cookies = cookie.load('userToken')
    const doAuth = async () => {
      dispatch({
        type: SET_AUTHENTICATING,
        isAuthenticating: true,
      })

      const res = await axios.get('http://localhost:3030/users/authenticate', {
        headers: {
          Authorization: `Bearer ${cookies}`,
        },
      })

      dispatch({
        type: SET_USER,
        user: res.data.user,
      })
      dispatch({
        type: SET_LOGIN_STATE,
        isLoggedIn: true,
      })
      dispatch({
        type: SET_AUTHENTICATING,
        isAuthenticating: false,
      })
    }
    if (typeof window !== undefined) {
      if (cookies) {
        doAuth()
      } else {
        cookie.remove('userToken')
        dispatch({
          type: SET_USER,
          user: null,
        })
        dispatch({
          type: SET_LOGIN_STATE,
          isLoggedIn: false,
        })
      }
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [])

  return (
    <AuthStateContext.Provider value={state}>
      <AuthDispatchContext.Provider value={dispatch}>{children}</AuthDispatchContext.Provider>
    </AuthStateContext.Provider>
  )
}

const useAuthState = () => {
  const context = React.useContext(AuthStateContext)

  return context
}

const useAuthDispatch = () => {
  const context = React.useContext(AuthDispatchContext)

  return context
}

export { AuthProvider, useAuthState, useAuthDispatch }
