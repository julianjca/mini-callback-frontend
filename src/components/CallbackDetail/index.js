import React from 'react'
import { Flex, Box, Button, Heading, Badge, Text, Center } from '@chakra-ui/react'
import { useRouter } from 'next/router'

import { useAuthDispatch } from '../../context/auth'
import { DO_LOGOUT } from '../../constants'

import { mapColorScheme } from '../../lib'

const CallbackDetail = ({ callback }) => {
  const router = useRouter()
  const dispatch = useAuthDispatch()

  return (
    <Center flexDirection="column" pt={20} pb={20}>
      <Flex width="100%" maxWidth={600} alignItems="center" justifyContent="space-between">
        <Box
          onClick={() => router.push('/dashboard', '/dashboard', { shallow: true })}
          cursor="pointer"
        >
          <Heading>Callback UI</Heading>
        </Box>
        <Box>
          <Button
            onClick={() => dispatch({ type: DO_LOGOUT })}
            size="lg"
            width={100}
            colorScheme="blue"
          >
            Log out
          </Button>
        </Box>
      </Flex>
      <Center width="100%" mt={50}>
        <Box maxWidth={600} width="100%" borderWidth="1px" borderRadius="lg" overflow="hidden">
          <Box p="6">
            <Box d="flex" alignItems="baseline">
              <Badge
                borderRadius="full"
                px="2"
                colorScheme={mapColorScheme(callback.callbackResponseCode)}
              >
                {callback.callbackResponseCode}
              </Badge>
              <Box
                color="gray.500"
                fontWeight="semibold"
                letterSpacing="wide"
                fontSize="xs"
                textTransform="uppercase"
                ml="2"
              >
                {callback.id}
              </Box>
            </Box>

            <Box
              display="flex"
              justifyContent="space-between"
              mt="5"
              fontWeight="semibold"
              as="h4"
              lineHeight="tight"
              isTruncated
            >
              Virtual Account:{' '}
              <Text as="span" display="inline-block" fontWeight="normal">
                {callback.virtual_account}
              </Text>
            </Box>
            <Box
              display="flex"
              justifyContent="space-between"
              mt="1"
              fontWeight="semibold"
              as="h4"
              lineHeight="tight"
              isTruncated
            >
              Bank Code:{' '}
              <Text as="span" display="inline-block" fontWeight="normal">
                {callback.bank_code}
              </Text>
            </Box>
            <Box
              display="flex"
              justifyContent="space-between"
              mt="1"
              fontWeight="semibold"
              as="h4"
              lineHeight="tight"
              isTruncated
            >
              Timestamp:{' '}
              <Text as="span" display="inline-block" fontWeight="normal">
                {callback.timestamp}
              </Text>
            </Box>
            <Box
              display="flex"
              justifyContent="space-between"
              mt="1"
              fontWeight="semibold"
              as="h4"
              lineHeight="tight"
              isTruncated
            >
              Transaction ID:{' '}
              <Text as="span" display="inline-block" fontWeight="normal">
                {callback.transaction_id}
              </Text>
            </Box>
            <Box
              display="flex"
              justifyContent="space-between"
              mt="1"
              fontWeight="semibold"
              as="h4"
              lineHeight="tight"
              isTruncated
            >
              Business Code:{' '}
              <Text as="span" display="inline-block" fontWeight="normal">
                {callback.businessId}
              </Text>
            </Box>
            <Box mt={5} textAlign="right">
              {callback.callbackResponseCode >= 300 && <Button>Retry</Button>}
            </Box>
          </Box>
        </Box>
      </Center>
    </Center>
  )
}

export default CallbackDetail
