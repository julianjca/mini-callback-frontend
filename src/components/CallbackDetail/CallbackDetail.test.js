/* eslint-disable testing-library/prefer-screen-queries */
import React from 'react'
import { render, fireEvent, waitFor } from '../../../test-utils'
import * as context from '../../context/auth'
import * as nextRouter from 'next/router'
import { DO_LOGOUT } from '../../constants'
import axios from 'axios'

import CallbackDetail from './index'

jest.mock('../../context/auth')
jest.mock('next/router')
jest.mock('axios')

// eslint-disable-next-line no-import-assign
nextRouter.useRouter = jest.fn()
const push = jest.fn()
nextRouter.useRouter.mockImplementation(() => ({ push }))

const useAuthDispatch = jest.fn()
context.useAuthDispatch.mockImplementation(() => useAuthDispatch)

const callback = {
  bank_code: 'BANK_ABC',
  businessId: '270cd61d-1965-4ed2-bbf3-a91425a030b8',
  callbackResponseCode: 200,
  createdAt: '2021-02-18T12:54:16.810Z',
  id: 'e8980029-04a0-4b9c-8c64-182c8a81cc43',
  timestamp: '2021-02-18T03:13:01.411Z',
  transaction_id: '123',
  updatedAt: '2021-02-18T12:54:16.810Z',
  virtual_account: '1234',
}

describe('CallbackDetail', () => {
  test('renders the right text', () => {
    const element = render(<CallbackDetail callback={callback} />)

    expect(element).toMatchSnapshot()
  })

  test('runs log out function', () => {
    const element = render(<CallbackDetail callback={callback} />)

    const logOutButton = element.getByText('Log out')
    fireEvent.click(logOutButton)
    expect(useAuthDispatch).toHaveBeenCalledTimes(1)
    expect(useAuthDispatch).toHaveBeenCalledWith({
      type: DO_LOGOUT,
    })
  })

  test('renders the correct callback data', () => {
    const element = render(<CallbackDetail callback={callback} />)

    expect(element.getByText('BANK_ABC')).toBeInTheDocument()
    expect(element.getByText('270cd61d-1965-4ed2-bbf3-a91425a030b8')).toBeInTheDocument()
    expect(element.getByText('200')).toBeInTheDocument()
    expect(element.getByText('e8980029-04a0-4b9c-8c64-182c8a81cc43')).toBeInTheDocument()
    expect(element.getByText('123')).toBeInTheDocument()
    expect(element.getByText('1234')).toBeInTheDocument()
  })

  test('able to retry', async () => {
    const element = render(<CallbackDetail callback={{ ...callback, callbackResponseCode: 400 }} />)

    expect(element.getByText('400')).toBeInTheDocument()

    const retryButton = element.getByText('Retry')
    expect(retryButton).toBeInTheDocument()

    const axiosMock = jest.spyOn(axios, 'put').mockImplementation(() =>
      Promise.resolve({
        data: {
          callback: { ...callback, callbackResponseCode: 200 },
        },
      }),
    )

    fireEvent.click(retryButton)

    await waitFor(() => expect(element.queryByText('Retry')).not.toBeInTheDocument())
    await waitFor(() => expect(element.getByText('200')).toBeInTheDocument())

    axiosMock.mockClear()
  })
})
