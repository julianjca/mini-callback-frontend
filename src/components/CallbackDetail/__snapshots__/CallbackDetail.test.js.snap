// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CallbackDetail renders the right text 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body
    class="chakra-ui-light"
  >
    <div
      id="chakra-toast-portal"
    >
      <ul
        id="chakra-toast-manager-top"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column; margin: 0px auto; top: 0px; right: 0px; left: 0px;"
      />
      <ul
        id="chakra-toast-manager-top-left"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column; top: 0px; left: 0px;"
      />
      <ul
        id="chakra-toast-manager-top-right"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column; top: 0px; right: 0px;"
      />
      <ul
        id="chakra-toast-manager-bottom-left"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column; bottom: 0px; left: 0px;"
      />
      <ul
        id="chakra-toast-manager-bottom"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column; margin: 0px auto; bottom: 0px; right: 0px; left: 0px;"
      />
      <ul
        id="chakra-toast-manager-bottom-right"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column; bottom: 0px; right: 0px;"
      />
    </div>
    <div>
      <div
        class="css-127d6hh"
      >
        <div
          class="css-eof6gw"
        >
          <div
            class="css-4g6ai3"
          >
            <h2
              class="chakra-heading css-zey6tx"
            >
              Callback UI
            </h2>
          </div>
          <div
            class="css-0"
          >
            <button
              class="chakra-button css-1tjlhec"
              type="button"
            >
              Log out
            </button>
          </div>
        </div>
        <div
          class="css-y82d7d"
        >
          <div
            class="css-6sz6xx"
          >
            <div
              class="css-am8j60"
            >
              <div
                class="css-x1sij0"
              >
                <span
                  class="chakra-badge css-tjtztc"
                >
                  200
                </span>
                <div
                  class="css-1ii03pb"
                >
                  e8980029-04a0-4b9c-8c64-182c8a81cc43
                </div>
              </div>
              <h4
                class="css-y9r156"
              >
                Virtual Account:
                 
                <span
                  class="chakra-text css-tn1jlw"
                >
                  1234
                </span>
              </h4>
              <h4
                class="css-88sfl1"
              >
                Bank Code:
                 
                <span
                  class="chakra-text css-tn1jlw"
                >
                  BANK_ABC
                </span>
              </h4>
              <h4
                class="css-88sfl1"
              >
                Timestamp:
                 
                <span
                  class="chakra-text css-tn1jlw"
                >
                  2021-02-18T03:13:01.411Z
                </span>
              </h4>
              <h4
                class="css-88sfl1"
              >
                Transaction ID:
                 
                <span
                  class="chakra-text css-tn1jlw"
                >
                  123
                </span>
              </h4>
              <h4
                class="css-88sfl1"
              >
                Business Code:
                 
                <span
                  class="chakra-text css-tn1jlw"
                >
                  270cd61d-1965-4ed2-bbf3-a91425a030b8
                </span>
              </h4>
              <div
                class="css-1ror4vt"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="css-127d6hh"
    >
      <div
        class="css-eof6gw"
      >
        <div
          class="css-4g6ai3"
        >
          <h2
            class="chakra-heading css-zey6tx"
          >
            Callback UI
          </h2>
        </div>
        <div
          class="css-0"
        >
          <button
            class="chakra-button css-1tjlhec"
            type="button"
          >
            Log out
          </button>
        </div>
      </div>
      <div
        class="css-y82d7d"
      >
        <div
          class="css-6sz6xx"
        >
          <div
            class="css-am8j60"
          >
            <div
              class="css-x1sij0"
            >
              <span
                class="chakra-badge css-tjtztc"
              >
                200
              </span>
              <div
                class="css-1ii03pb"
              >
                e8980029-04a0-4b9c-8c64-182c8a81cc43
              </div>
            </div>
            <h4
              class="css-y9r156"
            >
              Virtual Account:
               
              <span
                class="chakra-text css-tn1jlw"
              >
                1234
              </span>
            </h4>
            <h4
              class="css-88sfl1"
            >
              Bank Code:
               
              <span
                class="chakra-text css-tn1jlw"
              >
                BANK_ABC
              </span>
            </h4>
            <h4
              class="css-88sfl1"
            >
              Timestamp:
               
              <span
                class="chakra-text css-tn1jlw"
              >
                2021-02-18T03:13:01.411Z
              </span>
            </h4>
            <h4
              class="css-88sfl1"
            >
              Transaction ID:
               
              <span
                class="chakra-text css-tn1jlw"
              >
                123
              </span>
            </h4>
            <h4
              class="css-88sfl1"
            >
              Business Code:
               
              <span
                class="chakra-text css-tn1jlw"
              >
                270cd61d-1965-4ed2-bbf3-a91425a030b8
              </span>
            </h4>
            <div
              class="css-1ror4vt"
            />
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
